### PR TITLE
Minor edits to the Descriptor HowTo Guide

### PR DESCRIPTION
**Pull request by geryogam**
This PR provides the following changes to the [*Descriptor HowTo Guide*](https://docs.python.org/3/howto/descriptor.html):

- fix a typo;
- fix a Sphinx tag preventing syntax highlighting;
- fix a bug in `ClassMethod`;
- reuse an existing `cls` name in `Object`;
- allow keyword arguments in `Type.__new__` and `Object.__new__`;
- allow attribute lookup from a class in `Member.__get__`;
- use two more `super()` in `Object` and `Type` for consistency.

### CODE REVIEW

**Comment by merwok** [comment]
Can you open a ticket (with Raymond Hettinger in nosy) to discuss the problems you have identified here?

**Comment by maggyero** [comment]
@merwok Thanks for the suggestion, that’s what I intended to do. The rest of the PR is completed but I still have to finish the last section on `__slots__` (I completed my tests on it last year but I did not take the extra days necessary to update this PR). I will try to find a weekend shortly to do it.

File: Doc/howto/descriptor.rst [hunk 1]
```diff
@@ -795,7 +795,7 @@ afterwards, :meth:`__set_name__` will need to be called manually.
 ORM example
 -----------
 
-The following code is simplified skeleton showing how data descriptors could
+The following code is a simplified skeleton showing how data descriptors could
```
Comment (rhettinger) [inline Doc/howto/descriptor.rst:850]: This is correct.

File: Doc/howto/descriptor.rst [hunk 2]
```diff
@@ -940,6 +940,8 @@ here is a pure Python equivalent:
             self._name = name
 
         def __get__(self, obj, objtype=None):
+            if obj is None and objtype is None:
```
Comment (rhettinger) [inline Doc/howto/descriptor.rst]: This is not relevant to understanding properties.  The code is not meant to be a drop in substitute that covers all corner test cases.  The edit gets in the way of the essential message about what properties do.  Even the pure Python in equivalent in the comments section of the C code does not have this.

File: Doc/howto/descriptor.rst [hunk 3]
```diff
@@ -940,6 +940,8 @@ here is a pure Python equivalent:
...
+                raise TypeError('__get__(None, None) is invalid')
```
Comment (rhettinger) [inline Doc/howto/descriptor.rst]: same

File: Doc/howto/descriptor.rst [hunk 4]
```diff
@@ -1087,6 +1089,8 @@ during dotted lookup from an instance.  Here's how it works:
 
         def __get__(self, obj, objtype=None):
             "Simulate func_descr_get() in Objects/funcobject.c"
+            if obj is None and objtype is None:
```
Comment (rhettinger) [inline Doc/howto/descriptor.rst]: same

File: Doc/howto/descriptor.rst [hunk 5]
```diff
@@ -1087,6 +1089,8 @@ during dotted lookup from an instance.  Here's how it works:
...
+                raise TypeError('__get__(None, None) is invalid')
```
Comment (rhettinger) [inline Doc/howto/descriptor.rst]: same

File: Doc/howto/descriptor.rst [hunk 6]
```diff
@@ -1277,9 +1283,11 @@ Using the non-data descriptor protocol, a pure Python version of
             self.f = f
 
         def __get__(self, obj, cls=None):
+            if obj is None and cls is None:
+                raise TypeError('__get__(None, None) is invalid')
```
Comment (rhettinger) [inline Doc/howto/descriptor.rst]: same

File: Doc/howto/descriptor.rst [hunk 7]
```diff
@@ -1205,7 +1209,7 @@ example calls are unexciting:
 Using the non-data descriptor protocol, a pure Python version of
 :func:`staticmethod` would look like this:
 
-.. doctest::
+.. testcode::
```
Comment (rhettinger) [inline Doc/howto/descriptor.rst]: Yes, this is a correct, useful change.

File: Doc/howto/descriptor.rst [hunk 8]
```diff
@@ -1277,9 +1283,11 @@ Using the non-data descriptor protocol, a pure Python version of
             self.f = f
 
         def __get__(self, obj, cls=None):
+            if obj is None and cls is None:
+                raise TypeError('__get__(None, None) is invalid')
             if cls is None:
                 cls = type(obj)
-            if hasattr(obj, '__get__'):
+            if hasattr(self.f, '__get__'):
```
Comment (rhettinger) [inline Doc/howto/descriptor.rst]: Out-of-date.  This currently reads:
```

    class ClassMethod:
        "Emulate PyClassMethod_Type() in Objects/funcobject.c"

        def __init__(self, f):
            self.f = f

        def __get__(self, obj, cls=None):
            if cls is None:
                cls = type(obj)
            if hasattr(type(self.f), '__get__'):
                return self.f.__get__(cls, cls)
            return MethodType(self.f, cls)
```

File: Doc/howto/descriptor.rst [hunk 9]
```diff
@@ -1303,7 +1311,7 @@ Using the non-data descriptor protocol, a pure Python version of
     >>> t.cm(11, 22)
     (<class 'T'>, 11, 22)
 
-The code path for ``hasattr(obj, '__get__')`` was added in Python 3.9 and
+The code path for ``hasattr(self.f, '__get__')`` was added in Python 3.9 and
```
Comment (rhettinger) [inline Doc/howto/descriptor.rst]: Out-of-date.  The current text is:

`
The code path for ``hasattr(type(self.f), '__get__')`` was added in`

File: Doc/howto/descriptor.rst [hunk 10]
```diff
@@ -1413,7 +1421,7 @@ It is not possible to create an exact drop-in pure Python version of
 ``__slots__`` because it requires direct access to C structures and control
 over object memory allocation.  However, we can build a mostly faithful
 simulation where the actual C structure for slots is emulated by a private
-``_slotvalues`` list.  Reads and writes to that private structure are managed
+``_slot_values`` list.  Reads and writes to that private structure are managed
```
Comment (rhettinger) [inline Doc/howto/descriptor.rst]: Let's let this as is.  It is just a question of taste.  We normally only add a extra underscore if the combined word is hard to mentally separate.

File: Doc/howto/descriptor.rst [hunk 11]
```diff
@@ -1432,25 +1440,29 @@ by member descriptors:
         def __get__(self, obj, objtype=None):
             'Emulate member_get() in Objects/descrobject.c'
             # Also see PyMember_GetOne() in Python/structmember.c
-            value = obj._slotvalues[self.offset]
+            if obj is None and objtype is None:
+                raise TypeError('__get__(None, None) is invalid')
```
Comment (rhettinger) [inline Doc/howto/descriptor.rst]: Same comment as before.

File: Doc/howto/descriptor.rst [hunk 12]
```diff
@@ -1460,13 +1472,25 @@ variables:
     class Type(type):
         'Simulate how the type metaclass adds member objects for slots'
 
-        def __new__(mcls, clsname, bases, mapping):
+        def __new__(mcls, clsname, bases, mapping, **kwargs):
             'Emuluate type_new() in Objects/typeobject.c'
             # type_new() calls PyTypeReady() which calls add_methods()
             slot_names = mapping.get('slot_names', [])
-            for offset, name in enumerate(slot_names):
+            if slot_names:
```
Comment (rhettinger) [inline Doc/howto/descriptor.rst]: I don't like these edits (perhaps just a personal preference).  The existing code is much easier to match to and keep in sync with the C code (which is our objective here).  Also, this pure python equivalent has been tested on advanced python students and doesn't seem to have any issues with intelligibility beyond the intrinsic cyclomatic complexity.

File: Doc/howto/descriptor.rst [hunk 13]
```diff
@@ -1477,20 +1501,30 @@ Python:
     class Object:
         'Simulate how object.__new__() allocates memory for __slots__'
 
-        def __new__(cls, *args):
+        def __new__(cls, *args, **kwargs):
```
Comment (rhettinger) [inline Doc/howto/descriptor.rst:1585]: I'll review the rest of these later.   The title "fix the Python equivalents" implies that these we broken examples.    Many of these proposed edits are stylistic and exceed what the document was attempting to accomplish. 

The code existing is tested for fidelity to the C version and covers the essential points that we're trying to communicate.  There is no goal to make a drop-in replacement or to add code that distracts from the essentials.  User/reader testing shows that the existing code is already near the outer limits of what readers can absorb.

File: Doc/howto/descriptor.rst [hunk 14]
```diff
@@ -1477,20 +1501,30 @@ Python:
...
+                )
+            return super().__getattribute__(name)
+
         def __setattr__(self, name, value):
             'Emulate _PyObject_GenericSetAttrWithDict() Objects/object.c'
             cls = type(self)
             if hasattr(cls, 'slot_names') and name not in cls.slot_names:
                 raise AttributeError(
-                    f'{type(self).__name__!r} object has no attribute {name!r}'
+                    f'{cls.__name__!r} object has no attribute {name!r}'
```
Comment (rhettinger) [inline Doc/howto/descriptor.rst:1598]: I like this.

File: Doc/howto/descriptor.rst [hunk 15]
```diff
@@ -1499,7 +1533,7 @@ Python:
...
```
Comment (rhettinger) [inline Doc/howto/descriptor.rst:1607]: Same.

File: Doc/howto/descriptor.rst [hunk 16]
```diff
@@ -1533,31 +1567,44 @@ At this point, the metaclass has loaded member objects for *x* and *y*::
 
     # We test this separately because the preceding section is not
     # doctestable due to the hex memory address for the __init__ function
-    >>> isinstance(vars(H)['x'], Member)
+    >>> isinstance(H.x, Member)
```
Comment (rhettinger) [inline Doc/howto/descriptor.rst]: No thanks.  I'm specifically relating this code back to the listing of what is in the instance dict.

**Review by rhettinger — CHANGES_REQUESTED** [review]

File: Doc/howto/descriptor.rst (hunk 1 above)
Comment (geryogam) [reply Doc/howto/descriptor.rst:850]: Alright.

File: Doc/howto/descriptor.rst (hunk 2 above)
Comment (geryogam) [reply Doc/howto/descriptor.rst]: I agree that the introduced runtime type checks harms readability. I am removing them.

File: Doc/howto/descriptor.rst (hunk 4 above)
Comment (geryogam) [reply Doc/howto/descriptor.rst]: Okay.

File: Doc/howto/descriptor.rst (hunk 5 above)
Comment (geryogam) [reply Doc/howto/descriptor.rst]: Okay.

File: Doc/howto/descriptor.rst (hunk 6 above)
Comment (geryogam) [reply Doc/howto/descriptor.rst]: Okay.

File: Doc/howto/descriptor.rst (hunk 7 above)
Comment (geryogam) [reply Doc/howto/descriptor.rst]: Alright.

File: Doc/howto/descriptor.rst (hunk 8 above)
Comment (geryogam) [reply Doc/howto/descriptor.rst]: Alright. I wish we had a convenient way in Python to bypass instance attributes (with `type(self.f)`, the attribute `'__get__'` is also looked up in the metaclass). I suggested this on Stack Overflow: https://stackoverflow.com/q/67099920/2326961

File: Doc/howto/descriptor.rst (hunk 9 above)
Comment (maggyero) [reply Doc/howto/descriptor.rst:1314]: Okay.

Comment (geryogam) [reply Doc/howto/descriptor.rst]: Alright.

File: Doc/howto/descriptor.rst (hunk 10 above)
Comment (geryogam) [reply Doc/howto/descriptor.rst]: Alright.

File: Doc/howto/descriptor.rst (hunk 3 above)
Comment (geryogam) [reply Doc/howto/descriptor.rst]: Okay.

File: Doc/howto/descriptor.rst (hunk 11 above)
Comment (geryogam) [reply Doc/howto/descriptor.rst]: Okay.

File: Doc/howto/descriptor.rst (hunk 14 above)
Comment (geryogam) [reply Doc/howto/descriptor.rst:1598]: Alright.

File: Doc/howto/descriptor.rst (hunk 15 above)
Comment (geryogam) [reply Doc/howto/descriptor.rst:1607]: Okay.

File: Doc/howto/descriptor.rst (hunk 16 above)
Comment (geryogam) [reply Doc/howto/descriptor.rst]: Alright.

File: Doc/howto/descriptor.rst (hunk 12 above)
Comment (geryogam) [reply Doc/howto/descriptor.rst]: I agree.

File: Doc/howto/descriptor.rst (hunk 13 above)
Comment (geryogam) [reply Doc/howto/descriptor.rst:1585]: I agree, I have updated the title.

**Comment by maggyero** [comment]
Thanks @rhettinger, I have made the requested changes; please review again.

File: Doc/howto/descriptor.rst [hunk 17]
```diff
@@ -1560,20 +1562,20 @@ Python:
...
         'Simulate how object.__new__() allocates memory for __slots__'
 
-        def __new__(cls, *args):
+        def __new__(cls, *args, **kwargs):
             'Emulate object_new() in Objects/typeobject.c'
             inst = super().__new__(cls)
             if hasattr(cls, 'slot_names'):
                 empty_slots = [null] * len(cls.slot_names)
-                object.__setattr__(inst, '_slotvalues', empty_slots)
+                super(Object, inst).__setattr__('_slotvalues', empty_slots)
```
Comment (merwok) [inline Doc/howto/descriptor.rst]: I wonder why the code was not already using `super().__setattr__`, given that super *is* used three lines before.
Maybe the author really wanted `object.__setattr__`.

File: Doc/howto/descriptor.rst [hunk 18]
```diff
@@ -1543,13 +1545,13 @@ variables:
...
 
-        def __new__(mcls, clsname, bases, mapping):
+        def __new__(mcls, clsname, bases, mapping, **kwargs):
             'Emulate type_new() in Objects/typeobject.c'
             # type_new() calls PyTypeReady() which calls add_methods()
             slot_names = mapping.get('slot_names', [])
             for offset, name in enumerate(slot_names):
                 mapping[name] = Member(name, clsname, offset)
-            return type.__new__(mcls, clsname, bases, mapping)
+            return super().__new__(mcls, clsname, bases, mapping)
```
Comment (rhettinger) [inline Doc/howto/descriptor.rst]: Restore the `type.__new__` and add `**kwargs` to this line.

File: Doc/howto/descriptor.rst [hunk 19]
```diff
@@ -1515,6 +1515,8 @@ by member descriptors:
         def __get__(self, obj, objtype=None):
             'Emulate member_get() in Objects/descrobject.c'
             # Also see PyMember_GetOne() in Python/structmember.c
+            if obj is None:
+                return self
```
Comment (rhettinger) [inline Doc/howto/descriptor.rst:1539]: This is okay.

File: Doc/howto/descriptor.rst [hunk 20]
```diff
@@ -827,7 +827,7 @@ afterwards, :meth:`__set_name__` will need to be called manually.
 ORM example
 -----------
 
-The following code is simplified skeleton showing how data descriptors could
+The following code is a simplified skeleton showing how data descriptors could
```
Comment (rhettinger) [inline Doc/howto/descriptor.rst:850]: This is okay.

File: Doc/howto/descriptor.rst [hunk 21]
```diff
@@ -1560,20 +1562,20 @@ Python:
     class Object:
         'Simulate how object.__new__() allocates memory for __slots__'
 
-        def __new__(cls, *args):
+        def __new__(cls, *args, **kwargs):
```
Comment (rhettinger) [inline Doc/howto/descriptor.rst:1585]: This is okay but mostly irrelevant to the example.

File: Doc/howto/descriptor.rst (hunk 17 above)
Comment (rhettinger) [reply Doc/howto/descriptor.rst]: Let's let this as-is. The two argument form of `super()` is distracting and doesn't add value to the example.

File: Doc/howto/descriptor.rst [hunk 22]
```diff
@@ -1560,20 +1562,20 @@ Python:
...
+                super(Object, inst).__setattr__('_slotvalues', empty_slots)
             return inst
 
         def __setattr__(self, name, value):
             'Emulate _PyObject_GenericSetAttrWithDict() Objects/object.c'
             cls = type(self)
             if hasattr(cls, 'slot_names') and name not in cls.slot_names:
                 raise AttributeError(
-                    f'{type(self).__name__!r} object has no attribute {name!r}'
+                    f'{cls.__name__!r} object has no attribute {name!r}'
```
Comment (rhettinger) [inline Doc/howto/descriptor.rst:1598]: This is okay.

File: Doc/howto/descriptor.rst [hunk 23]
```diff
@@ -1582,7 +1584,7 @@ Python:
...
```
Comment (rhettinger) [inline Doc/howto/descriptor.rst:1607]: This is okay.

**Review by rhettinger — CHANGES_REQUESTED** [review]

File: Doc/howto/descriptor.rst (hunk 18 above)
Comment (geryogam) [reply Doc/howto/descriptor.rst]: Updated.

File: Doc/howto/descriptor.rst (hunk 17 above)
Comment (geryogam) [reply Doc/howto/descriptor.rst]: Updated.

**Comment by merwok** [comment]
Please open a ticket to track this properly.

File: Doc/howto/descriptor.rst [hunk 24]
```diff
@@ -1580,9 +1582,9 @@ Python:
     class Object:
         'Simulate how object.__new__() allocates memory for __slots__'
 
-        def __new__(cls, *args):
+        def __new__(cls, *args, **kwargs):
```
Comment (rhettinger) [inline Doc/howto/descriptor.rst:1585]: Please remove this edit.

File: Doc/howto/descriptor.rst [hunk 25]
```diff
@@ -1580,9 +1582,9 @@ Python:
...
             'Emulate object_new() in Objects/typeobject.c'
-            inst = super().__new__(cls)
+            inst = super().__new__(cls, *args, **kwargs)
```
Comment (rhettinger) [inline Doc/howto/descriptor.rst]: Remove this edit.  It is a breaking change.   

```
Traceback (most recent call last):
  File "/Users/raymond/Documents/tmp20.py", line 84, in <module>
    h = H(10, 20)
  File "/Users/raymond/Documents/tmp20.py", line 51, in __new__
    inst = super().__new__(cls, *args, **kwargs)
TypeError: object.__new__() takes exactly one argument (the type to instantiate)

```

**Review by rhettinger — CHANGES_REQUESTED** [review]

File: Doc/howto/descriptor.rst (hunk 25 above)
Comment (geryogam) [reply Doc/howto/descriptor.rst]: Fixed.

**Comment by maggyero** [comment]
Thanks for the review.

**Pull request merged**